### PR TITLE
gh-115874: Fix segfault in `FutureIter_dealloc`

### PR DESCRIPTION
**Pull request by savannahostrowski**
Per discussion here: https://github.com/python/cpython/issues/115874#issuecomment-1967284036

* Issue: gh-115874

### CODE REVIEW

**Comment by savannahostrowski** [comment]
cc: @brandtbucher for his review as well

**Review by gvanrossum — COMMENTED** [review]
I'll leave it to @brandtbucher to review this -- I am somewhat disappointed that in order to fix this we have to copy a bunch of subtle code from inside `PyType_GetModuleByDef()` -- especially since there the code is enclosed in macros `BEGIN_TYPE_LOCK()` and `END_TYPE_LOCK()` (though those seem needed only because of the call to `lookup_tp_mro()` there).

At the same time I understand you don't want to call `PyType_GetModuleByDef()` and have to deal with the exception it raises in exactly the scenario we're trying to tiptoe around here.

**Comment by brandtbucher** [comment]
> At the same time I understand you don't want to call `PyType_GetModuleByDef()` and have to deal with the exception it raises in exactly the scenario we're trying to tiptoe around here.

It looks like it's actually a bit more subtle than that.

If I understand correctly, this isn't being done to avoid an exception, but to avoid a crash due to the type's MRO being `NULL` (this happens when the type and instance are part of a cycle, and the type is cleared before the instance is deallocated). In general, it seems unsafe to assume types aren't cleared in an instance's `tp_dealloc`, as we do in the current code.

**Comment by brandtbucher** [comment]
@savannahostrowski, do you mind adding a NEWS blurb? No need to get too technical, just explaining that we've fixed a possible crash during garbage-collection of `_asyncio.FutureIter` objects.

Can you also add a comment or two to the new code referencing the issue number and explaining that:
- We can't use `PyType_GetModuleByDef`, since the type might have already been cleared. This is also why we need to check that `ht_module` isn't `NULL`.
- Since it's uncommon to subclass this type, it's fine (and probably faster for the common case!) to just check our type's module and not bother walking the MRO.
- This means subclasses can't make use of the free list... oh well.

Can you also confirm for me that our old reproducer crashes on 3.12? If so, we can add the test (it's okay if it no longer works on 3.13, probably still good to have) and flag this for backport.

**Comment by gvanrossum** [comment]
> If I understand correctly, this isn't being done to avoid an exception, but to avoid a crash due to the type's MRO being `NULL` (this happens when the type and instance are part of a cycle, and the type is cleared before the instance is deallocated). In general, it seems unsafe to assume types aren't cleared in an instance's `tp_dealloc`, as we do in the current code.

Hm, that would point to a pretty universal problem (the instance being cleared after the type, when both are involved in a cycle). Why isn’t that crashing other code? What’s special about this example?

(I am pushing back on this because the fix requires breaking through a nice abstraction, potentially in many more cases.)

**Comment by savannahostrowski** [comment]
Thanks for the feedback, folks. I'll add some comments to the code.

@brandtbucher I can confirm that the segfault still repros on 3.12. I know you did some investigation in other modules and that's how you found this issue. That said, is it worth doing a bit more spelunking to understand if this is happening in other places as well? I'm too new to know how prolific this might be 😅 but I want to address concerns here.

**Comment by erlend-aasland** [comment]
@brandtbucher, @savannahostrowski: how about storing the state pointer in the instance struct? We do this for module state access in the `sqlite3.Connection` class (and a few other places in `Modules/`. We know that the state pointer is still valid, since the type object holds a strong reference to the module.

**Comment by brandtbucher** [comment]
> Hm, that would point to a pretty universal problem (the instance being cleared after the type, when both are involved in a cycle). Why isn’t that crashing other code? What’s special about this example?

What's special is that *only* dealloc functions are susceptible to this fragile situation. Doing anything other than untracking, clearing, and freeing objects in dealloc functions is rare. Clever stuff belongs in finalizers, where you have an object in a known sane state... the obvious exception being freelists like this, which *must* go in a dealloc function, since it's "freeing" the memory.

> (I am pushing back on this because the fix requires breaking through a nice abstraction, potentially in many more cases.)

In the original issue, we found a similar crash in `itertools.tee`. I was worried too, but after a quick scan of literally every function matching `r"\w+dealloc\("` in C code, this was the only other one that worried me.

A perhaps more "principled" fix would be to change `PyType_GetModuleByDef` to raise in the case where `tp_mro` is `NULL` and change all of its callers to handle the error instead of asserting that its return value is non-NULL (which is super common, likely copy-pasted around). But that's a lot more invasive, and will slow down the happy path of common operations to work around a really tricky edge case that doesn't affect that much code.

In my opinion, the bug here is assuming that your type hasn't been cleared in a dealloc func. That's incorrect.

**Comment by brandtbucher** [comment]
> I know you did some investigation in other modules and that's how you found this issue. That said, is it worth doing a bit more spelunking to understand if this is happening in other places as well? I'm new to this part of the codebase to really understand how prolific this might be 😅 but I want to address concerns here.

See my comment above. I'm reasonably confident that there aren't other offenders, but another pair of eyes definitely wouldn't hurt.

File: Modules/_asynciomodule.c [hunk 1]
```diff
@@ -1601,11 +1601,24 @@ static void
 FutureIter_dealloc(futureiterobject *it)
 {
     PyTypeObject *tp = Py_TYPE(it);
-    asyncio_state *state = get_asyncio_state_by_def((PyObject *)it);
+
+    assert(_PyType_HasFeature((PyTypeObject *)tp, Py_TPFLAGS_HEAPTYPE));
```
Comment (brandtbucher) [inline Modules/_asynciomodule.c]: Unnecessary cast:

```suggestion
    assert(_PyType_HasFeature(tp, Py_TPFLAGS_HEAPTYPE));
```

File: Modules/_asynciomodule.c [hunk 2]
```diff
@@ -1601,11 +1601,24 @@ static void
...
+
+    PyHeapTypeObject *ht = (PyHeapTypeObject*)tp;
+    PyObject *module = ht->ht_module;
```
Comment (brandtbucher) [inline Modules/_asynciomodule.c]: Minor suggestion, we can make this a one-liner. Your call for which is more readable:

```suggestion
    PyObject *module = ((PyHeapTypeObject*)tp)->ht_module;
```

File: Modules/_asynciomodule.c (hunk 1 above)
Comment (brandtbucher) [inline Modules/_asynciomodule.c]: Maybe add a comment explaining this assert (since `FutureIter` is a heap type, any subclasses must also be).

File: Modules/_asynciomodule.c [hunk 3]
```diff
@@ -1601,11 +1601,24 @@ static void
...
+    asyncio_state *state = NULL;
+
     PyObject_GC_UnTrack(it);
     tp->tp_clear((PyObject *)it);
 
-    if (state->fi_freelist_len < FI_FREELIST_MAXLEN) {
+    // GH-115874: We can't use PyType_GetModuleByDef here as the type might have
+    // already been cleared. This is also why we must check if ht_module != NULL.
+    // Subclasses also cannot make use of the free list.
```
Comment (brandtbucher) [inline Modules/_asynciomodule.c]: Maybe reword this sentence to better capture the causal relationship here (they can't use the free list *because* of this restriction).

And, more pedantically, it's subclasses that belong to a different module that can't use it.

File: Modules/_asynciomodule.c [hunk 4]
```diff
@@ -1601,11 +1601,24 @@ static void
...
+    if (module && _PyModule_GetDef(module) == &_asynciomodule) {
```
Comment (brandtbucher) [inline Modules/_asynciomodule.c:1618]: Another thought: the control-flow here might be a bit clearer if rewritten. I'm not sure if the early return makes things more or less complex, though, so I'll let you make the call.

```c
    if (module && _PyModule_GetDef(module) == &_asynciomodule) {
        state = get_asyncio_state(module);
        if (state->fi_freelist_len < FI_FREELIST_MAXLEN) {
            state->fi_freelist_len++;
            it->future = (FutureObj *)state->fi_freelist;
            state->fi_freelist = it;
            return;
        }
    }
    PyObject_GC_Del(it);
    Py_DECREF(tp);
}
```

(GH won't let me make suggestions outside of the diff.)

**Comment by gvanrossum** [comment]
>In my opinion, the bug here is assuming that your type hasn't been cleared in a clear or dealloc func. That's incorrect.

Hmm... But it's an easy trap to fall into. Not exactly a bug magnet (only two instances in CPython itself), perhaps, but hard to debug, and hard to reason about: it seems it only happens when the type is cleared first. So isn't the bug that the type is cleared while it still has instances?

Do we understand how exactly this happened? Is `module_clear` called prematurely?

**Comment by brandtbucher** [comment]
> So isn't the bug that the type is cleared while it still has instances?
> 
> Do we understand how exactly this happened? Is `module_clear` called prematurely?

My hunch is that we have a cycle: `instance -> type -> module -> [...] -> instance`. I don't think the GC can be faulted for breaking the cycle at `type -> module` in this case, since it just chooses an element to clear essentially at random. Here, it just happens to be the type.

**Comment by gvanrossum** [comment]
> My hunch is that we have a cycle: `instance -> type -> module -> [...] -> instance`. I don't think the GC can be faulted for breaking the cycle at `type -> module` in this case, since it just chooses an element to clear essentially at random. Here, it just happens to be the type.

Yeah, that's fair, we don't want to have to educate the GC about types and instances. I had further suspicions but none of them check out, and this is the simplest explanation.

Maybe we just need to add a warning to the docs for `PyModule_GetState` that it's not safe to call from a `tp_dealloc` handler?

**Comment by savannahostrowski** [comment]
@erlend-aasland I took a look at your suggestion but in this case, even if we stored the state in `futureiterobject`, I think that the state could still be cleared before the instance is deallocated, making it possibly as unreliable as the module to grab the necessary info for deallocation.

**Comment by erlend-aasland** [comment]
> @erlend-aasland I took a look at your suggestion but in this case, even if we stored the state in `futureiterobject`, I think that the state could still be cleared before the instance is deallocated, making it possibly as unreliable as the module to grab the necessary info for deallocation.

Yes, I noticed that after reading through the discussion again (which is why I marked my comment as outdated); it is an unfortunate issue. Well, thanks a lot for implementing this workaround :)

**Review by erlend-aasland — APPROVED** [review]

File: Modules/_asynciomodule.c (hunk 3 above)
Comment (gvanrossum) [reply Modules/_asynciomodule.c]: I agree, this line threw me off (I was expecting an "exact" type check on the next line, but no).

The rest LGTM!

File: Modules/_asynciomodule.c (hunk 4 above)
Comment (savannahostrowski) [reply Modules/_asynciomodule.c:1618]: Personally, I probably think the way it's written now is a bit easier to parse so I'll leave it as is unless someone else feels strongly.

**Comment by vstinner** [comment]
Thanks for fixing this crash!

File: Modules/_asynciomodule.c [hunk 5]
```diff
@@ -1601,11 +1601,25 @@ static void
 FutureIter_dealloc(futureiterobject *it)
 {
     PyTypeObject *tp = Py_TYPE(it);
-    asyncio_state *state = get_asyncio_state_by_def((PyObject *)it);
+
+    // FutureIter is a heap type so any subclass must also be a heap type.
```
Comment (kumaraditya303) [inline Modules/_asynciomodule.c:1605]: This comment is misleading, `FutureIter` is not a basetype so cannot have subclasses.
cc @savannahostrowski

File: Modules/_asynciomodule.c [hunk 6]
```diff
@@ -1601,11 +1601,25 @@ static void
...
+    PyObject *module = ((PyHeapTypeObject*)tp)->ht_module;
+    asyncio_state *state = NULL;
+
     PyObject_GC_UnTrack(it);
     tp->tp_clear((PyObject *)it);
 
-    if (state->fi_freelist_len < FI_FREELIST_MAXLEN) {
+    // GH-115874: We can't use PyType_GetModuleByDef here as the type might have
+    // already been cleared, which is also why we must check if ht_module != NULL.
+    // Due to this restriction, subclasses that belong to a different module
```
Comment (kumaraditya303) [inline Modules/_asynciomodule.c:1616]: Same

File: Misc/NEWS.d/next/Core and Builtins/2024-04-13-18-59-25.gh-issue-115874.c3xG-E.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Fixed a possible segfault during garbage collection of ``_asyncio.FutureIter`` objects
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-13-18-59-25.gh-issue-115874.c3xG-E.rst:1]: Minor: This should have been in Library section rather than core builtins, its a bug of asyncio library.

Comment (savannahostrowski) [reply Misc/NEWS.d/next/Core and Builtins/2024-04-13-18-59-25.gh-issue-115874.c3xG-E.rst:1]: I've opened a PR to address these comments in #121638
